src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -170,7 +171,13 @@ String getClassNameInternal() {
         cacheKey = new CityHash128Key(sql);
 
         // Parse or fetch SQL metadata from cache.
-        ParsedSQLCacheItem parsedSql = getOrCreateCachedParsedSQL(cacheKey, sql);
+        ParsedSQLCacheItem parsedSql = getCachedParsedSQL(cacheKey, sql);
+        if (null != parsedSql) {
+            isExecutedAtLeastOnce = true;
+        }
+        else {
+            parsedSql = parseAndCacheSQL(cacheKey, sql);
+        }
```

Comment:
Thanks for the analysis! 

I will take a deeper look into this. This statement doesn't make sense to me.
> Again, because no PreparedStatement instance is ever executed more than once, there are no handles to be shared. 

That is why we have the cache and move directly to sp_prepexec if we see that another instance has already done sp_executsql before. This should make the new instance re-use the state from the previous instance and create the handle. Subsequent instance should re-use the handle. 
```
// Because sp_executesql was already called on this SQL-text use
// regular prep/exec pattern.
if (cachedPreparedStatementHandle.hasExecutedAtLeastOnce())
	isExecutedAtLeastOnce = true;
```

This may simply be an indication that the default cache size is of 10 is too small.